### ISSUE TITLE
Poor type inference of generators when used with coroutine objects

### ISSUE THREAD

**Issue by AlanBogarin** [microsoft/pyright#10278]
## Environment data

-   Pylance version: 2025.4.1
-   OS and version: Windows 11
-   Python version (& distribution if applicable, e.g. Anaconda): 3.13.2

## Code Snippet

```python
import types
import typing
import asyncio


@types.coroutine
# shows: typing.AwaitableGenerator[Any, Unknown, None, Any]
# expected: typing.Generator[Any, None, None] or AwaitableGenerator[Any, None, None, Any]
def generator_based_coroutine():
    result = yield from asyncio.sleep(1)  # error: reportUnknownVariableType
    typing.reveal_type(result)  # error: shows Unknown, (expected: None)


class AwaitableImplementation:
    # pylance shows: Generator[Any, Unknown, None]
    def __await__(self):  # error: reportUnknownParameterType
        result = yield from asyncio.sleep(1, 0).__await__()  # ok
        typing.reveal_type(result)  # partial ok: int (Literal[0])


async def main():
    await generator_based_coroutine()  # ok: () -> Awaitable[None]
    await AwaitableImplementation()  # ok

asyncio.run(main())
```

## Repro Steps

1. select type checking mode: `strict`

## Expected behavior

I expected that the types for the `generator_based_coroutine` and `AwaitableImplementation.__await__` functions could be inferred. Currently the types that would be for `SendT` of `Generator[YieldT, SendT, ReturnT]` and `AwaitableGenerator[YieldT, SendT, ReturnT, _S]` are `Unknown`.

I just noticed that `typing.AwaitableGenerator` seems to have an unused type variable `_S` in [typeshed](https://github.com/python/typeshed/blob/2b7019b933ac8339e4ae8f4619a2a746446a8532/stdlib/typing.pyi#L538)

## Actual behavior

- `generator_based_coroutine` has type `() -> AwaitableGenerator[Any, Unknown, None, Any]`

  ![Image](https://github.com/user-attachments/assets/f5274551-904f-4ae0-a1ab-9d894d9d2a37)

- `AwaitableImplementation.__await__` has type `() -> Generator[Any, Unknown, None]`

  ![Image](https://github.com/user-attachments/assets/5b0885c0-f442-44b6-a30d-4327358fd5a0)

## Logs

<details>

<summary>Python Runtime</summary>

```
Runtime type is 'NoneType'
Runtime type is 'int'
```

</details>
<details>

<summary>VSCode Pylance Logs</summary>

```
[Info] (4456) Execution environment: 3.13.2 (global)
[Info] (4456)   Extra paths:
[Info] (4456)     (none)
[Info] (4456)   Python version: 3.13.2.final.0
[Info] (4456)   Python platform: Windows
[Info] (4456)   Search paths:
[Info] (4456)     c:\Users\Dionisia\.vscode\extensions\ms-python.vscode-pylance-2025.4.1\dist\typeshed-fallback\stdlib
[Info] (4456)     d:\Alan\code\projects\awaitable-tools
[Info] (4456)     d:\Alan\code\projects\awaitable-tools\typings
[Info] (4456)     c:\Users\Dionisia\.vscode\extensions\ms-python.vscode-pylance-2025.4.1\dist\typeshed-fallback\stubs\...
[Info] (4456)     c:\Users\Dionisia\.vscode\extensions\ms-python.vscode-pylance-2025.4.1\dist\bundled\stubs
[Info] (4456)     c:\Users\Dionisia\AppData\Local\Programs\Python\Python313\DLLs
[Info] (4456)     c:\Users\Dionisia\AppData\Local\Programs\Python\Python313\Lib
[Info] (4456)     c:\Users\Dionisia\AppData\Local\Programs\Python\Python313
[Info] (4456)     c:\Users\Dionisia\AppData\Local\Programs\Python\Python313\Lib\site-packages
[Info] (4456) Adding fs watcher for library directories:
 file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib
file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313
[Info] (4456) Adding fs watcher for directories:
 file:///d%3A/Alan/code/projects/awaitable-tools
[Log] (4456) Searching for source files
[Info] (4456) Found 6 source files
[Log] (4456) Adding fs watcher for files:
 file:///d%3A/Alan/code/projects/awaitable-tools/requirements.txt
file:///d%3A/Alan/code/projects/awaitable-tools/pyproject.toml
[Info] (4456) Execution environment: 3.13.2 (global)
[Info] (4456)   Extra paths:
[Info] (4456)     (none)
[Info] (4456)   Python version: 3.13.2.final.0
[Info] (4456)   Python platform: Windows
[Info] (4456)   Search paths:
[Info] (4456)     c:\Users\Dionisia\.vscode\extensions\ms-python.vscode-pylance-2025.4.1\dist\typeshed-fallback\stdlib
[Info] (4456)     \<default workspace root>
[Info] (4456)     \<default workspace root>\typings
[Info] (4456)     c:\Users\Dionisia\.vscode\extensions\ms-python.vscode-pylance-2025.4.1\dist\typeshed-fallback\stubs\...
[Info] (4456)     c:\Users\Dionisia\.vscode\extensions\ms-python.vscode-pylance-2025.4.1\dist\bundled\stubs
[Info] (4456)     c:\Users\Dionisia\AppData\Local\Programs\Python\Python313\DLLs
[Info] (4456)     c:\Users\Dionisia\AppData\Local\Programs\Python\Python313\Lib
[Info] (4456)     c:\Users\Dionisia\AppData\Local\Programs\Python\Python313
[Info] (4456)     c:\Users\Dionisia\AppData\Local\Programs\Python\Python313\Lib\site-packages
[Info] (4456) Adding fs watcher for library directories:
 file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib
file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313
[Info] (4456) Adding fs watcher for directories:
 file:///%3Cdefault%20workspace%20root%3E
[Log] (4456) Searching for source files
[Error] (4456) File or directory "\<default workspace root>" does not exist.
[Info] (4456) No source files found.
[Log] (4456) Adding fs watcher for files:
 file:///%3Cdefault%20workspace%20root%3E/requirements.txt
file:///%3Cdefault%20workspace%20root%3E/pyproject.toml
[Log] (4456) Server settings returned for workspace: file:///d%3A/Alan/code/projects/awaitable-tools: {
  "autoSearchPaths": true,
  "disableLanguageServices": false,
  "openFilesOnly": true,
  "useLibraryCodeForTypes": true,
  "watchForSourceChanges": true,
  "watchForLibraryChanges": true,
  "watchForConfigChanges": true,
  "typeCheckingMode": "strict",
  "diagnosticSeverityOverrides": {
    "reportShadowedImports": "warning"
  },
  "diagnosticBooleanOverrides": {
    "strictListInference": false,
    "strictDictionaryInference": false,
    "strictSetInference": false,
    "analyzeUnannotatedFunctions": true,
    "strictParameterNoneValue": true,
    "enableTypeIgnoreComments": true,
    "deprecateTypingAliases": false,
    "enableReachabilityAnalysis": false,
    "enableExperimentalFeatures": false,
    "disableBytesTypePromotions": false
  },
  "logLevel": "log",
  "autoImportCompletions": false,
  "indexing": true,
  "completeFunctionParens": false,
  "enableExtractCodeAction": true,
  "indexOptions": {
    "packageDepths": [
      [
        "sklearn",
        2,
        false
      ],
      [
        "matplotlib",
        2,
        false
      ],
      [
        "scipy",
        2,
        false
      ],
      [
        "django",
        2,
        false
      ],
      [
        "flask",
        2,
        false
      ],
      [
        "fastapi",
        2,
        false
      ]
    ],
    "regenerateStdLibIndices": false,
    "localStorageUri": {
      "_key": "d:\\alan\\code\\projects\\awaitable-tools",
      "_filePath": "d:\\Alan\\code\\projects\\awaitable-tools",
      "_query": "",
      "_fragment": "",
      "_originalString": "file:///d%3A/Alan/code/projects/awaitable-tools",
      "_isCaseSensitive": false,
      "_formattedString": "file:///d%3A/Alan/code/projects/awaitable-tools",
      "_normalizedPath": "d:/Alan/code/projects/awaitable-tools"
    },
    "includeAliasesFromUserFiles": false,
    "userFileIndexingLimit": 2000
  },
  "variableInlayTypeHints": false,
  "callArgumentNameInlayHints": "off",
  "functionReturnInlayTypeHints": false,
  "pytestParametersInlayTypeHints": false,
  "importFormat": "absolute",
  "includeFileSpecs": [],
  "excludeFileSpecs": [],
  "ignoreFileSpecs": [],
  "formatOnType": true,
  "taskListTokens": [],
  "enablePytestSupport": true,
  "gotoDefinitionInStringLiteral": true,
  "remapDiagnostics": false,
  "intelliCodeEnabled": true,
  "aiCodeActions": {
    "implementAbstractClasses": true
  },
  "generateWithTypeAnnotation": true,
  "languageServerMode": "default",
  "enableEditableInstalls": false,
  "enablePerfTelemetry": false,
  "enableCompactContext": false,
  "pythonPath": {
    "_key": "c:\\users\\dionisia\\appdata\\local\\programs\\python\\python313\\python.exe",
    "_filePath": "C:\\Users\\Dionisia\\AppData\\Local\\Programs\\Python\\Python313\\python.exe",
    "_query": "",
    "_fragment": "",
    "_isCaseSensitive": false
  },
  "pythonEnvironmentName": "3.13.2 (global)",
  "disableTaggedHints": false,
  "fixAll": [],
  "showOnlyDirectDependenciesInAutoImport": false,
  "extraCommitChars": false,
  "autoFormatStrings": false,
  "autoIndent": true,
  "autoSplitStrings": true,
  "supportRestructuredText": true,
  "cacheLSPData": false,
  "reportExtraTelemetry": false,
  "supportDocstringTemplate": false,
  "displayEnglishDiagnostics": false,
  "diagnosticsSource": "pylance",
  "pyrightVersion": "",
  "nodeExecutable": "",
  "editorIndentConfig": {
    "defaultTabSequence": "    ",
    "tabSize": 4
  }
}
[Log] (4456) Server settings returned for workspace: undefined: {
  "autoSearchPaths": true,
  "disableLanguageServices": false,
  "openFilesOnly": true,
  "useLibraryCodeForTypes": true,
  "watchForSourceChanges": true,
  "watchForLibraryChanges": true,
  "watchForConfigChanges": true,
  "typeCheckingMode": "strict",
  "diagnosticSeverityOverrides": {
    "reportShadowedImports": "warning"
  },
  "diagnosticBooleanOverrides": {
    "strictListInference": false,
    "strictDictionaryInference": false,
    "strictSetInference": false,
    "analyzeUnannotatedFunctions": true,
    "strictParameterNoneValue": true,
    "enableTypeIgnoreComments": true,
    "deprecateTypingAliases": false,
    "enableReachabilityAnalysis": false,
    "enableExperimentalFeatures": false,
    "disableBytesTypePromotions": false
  },
  "logLevel": "log",
  "autoImportCompletions": false,
  "indexing": false,
  "completeFunctionParens": false,
  "enableExtractCodeAction": true,
  "indexOptions": {
    "packageDepths": [
      [
        "sklearn",
        2,
        false
      ],
      [
        "matplotlib",
        2,
        false
      ],
      [
        "scipy",
        2,
        false
      ],
      [
        "django",
        2,
        false
      ],
      [
        "flask",
        2,
        false
      ],
      [
        "fastapi",
        2,
        false
      ]
    ],
    "regenerateStdLibIndices": false,
    "includeAliasesFromUserFiles": false,
    "userFileIndexingLimit": 2000
  },
  "variableInlayTypeHints": false,
  "callArgumentNameInlayHints": "off",
  "functionReturnInlayTypeHints": false,
  "pytestParametersInlayTypeHints": false,
  "importFormat": "absolute",
  "includeFileSpecs": [],
  "excludeFileSpecs": [],
  "ignoreFileSpecs": [],
  "formatOnType": true,
  "taskListTokens": [],
  "enablePytestSupport": true,
  "gotoDefinitionInStringLiteral": true,
  "remapDiagnostics": false,
  "intelliCodeEnabled": true,
  "aiCodeActions": {
    "implementAbstractClasses": true
  },
  "generateWithTypeAnnotation": true,
  "languageServerMode": "default",
  "enableEditableInstalls": false,
  "enablePerfTelemetry": false,
  "enableCompactContext": false,
  "pythonPath": {
    "_key": "c:\\users\\dionisia\\appdata\\local\\programs\\python\\python313\\python.exe",
    "_filePath": "c:\\Users\\Dionisia\\AppData\\Local\\Programs\\Python\\Python313\\python.exe",
    "_query": "",
    "_fragment": "",
    "_originalString": "file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/python.exe",
    "_isCaseSensitive": false,
    "_formattedString": "file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/python.exe"
  },
  "pythonEnvironmentName": "3.13.2 (global)",
  "disableTaggedHints": false,
  "fixAll": [],
  "showOnlyDirectDependenciesInAutoImport": false,
  "extraCommitChars": false,
  "autoFormatStrings": false,
  "autoIndent": true,
  "autoSplitStrings": true,
  "supportRestructuredText": true,
  "cacheLSPData": false,
  "reportExtraTelemetry": false,
  "supportDocstringTemplate": false,
  "displayEnglishDiagnostics": false,
  "diagnosticsSource": "pylance",
  "pyrightVersion": "",
  "nodeExecutable": "",
  "editorIndentConfig": {
    "defaultTabSequence": "    ",
    "tabSize": 4
  }
}
[Log] (4456) BG: Attempting to resolve using local imports: __builtins__
[Log] (4456) BG: [BG(1)] analyzing: file:///d%3A/Alan/code/projects/awaitable-tools/z.py ...
[Log] (4456) BG: [BG(1)]   parsing: file:///d%3A/Alan/code/projects/awaitable-tools/z.py (204ms)
[Log] (4456) BG: [BG(1)]   parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/builtins.pyi [fs read 1ms] (75ms)
[Log] (4456) BG: [BG(1)]   binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/builtins.pyi (24ms)
[Log] (4456) BG: [BG(1)]   binding: file:///d%3A/Alan/code/projects/awaitable-tools/z.py (0ms)
[Log] (4456) BG: [BG(1)]   checking: file:///d%3A/Alan/code/projects/awaitable-tools/z.py ...
[Log] (4456) BG: [BG(1)]     parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/typing.pyi [fs read 3ms] (21ms)
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/typing.pyi (10ms)
[Log] (4456) BG: [BG(1)]     parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/types.pyi [fs read 1ms] (11ms)
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/types.pyi (5ms)
[Log] (4456) BG: [BG(1)]     parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi [fs read 1ms] (4ms)
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/_typeshed/__init__.pyi (2ms)
[Log] (4456) BG: [BG(1)]     parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/abc.pyi [fs read 1ms] (2ms)
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/abc.pyi (0ms)
[Log] (4456) BG: [BG(1)]     parsing: file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib/types.py [fs read 0ms] (4ms)
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib/types.py (4ms)
[Log] (4456) BG: [BG(1)]     parsing: file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib/typing.py [fs read 2ms] (37ms)
[Info] (4456) BG: Could not import '_typing' in file 'c:\Users\Dionisia\AppData\Local\Programs\Python\Python313\Lib\typing.py'
[Info] (4456) BG:   Looking in stubPath 'file:///d%3A/Alan/code/projects/awaitable-tools/typings'
[Info] (4456) BG:   Attempting to resolve stub package using root path 'file:///d%3A/Alan/code/projects/awaitable-tools/typings'
[Info] (4456) BG:   Attempting to resolve using root path 'file:///d%3A/Alan/code/projects/awaitable-tools/typings'
[Info] (4456) BG:   Looking in root directory of execution environment 'file:///d%3A/Alan/code/projects/awaitable-tools'
[Info] (4456) BG:   Attempting to resolve stub package using root path 'file:///d%3A/Alan/code/projects/awaitable-tools'
[Info] (4456) BG:   Attempting to resolve using root path 'file:///d%3A/Alan/code/projects/awaitable-tools'
[Info] (4456) BG:   Looking for typeshed stdlib path
[Info] (4456) BG:   Looking for typeshed stdlib path
[Info] (4456) BG:   Attempting to resolve using root path 'file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib'
[Info] (4456) BG:   Typeshed path not found
[Info] (4456) BG:   Finding python search paths
[Info] (4456) BG:   Executing interpreter: 'c:\Users\Dionisia\AppData\Local\Programs\Python\Python313\python.exe'
[Info] (4456) BG:   Skipping 'c:\Users\Dionisia\AppData\Local\Programs\Python\Python313\python313.zip' because it is not a valid directory
[Info] (4456) BG:   Received 4 paths from interpreter
[Info] (4456) BG:     file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/DLLs
[Info] (4456) BG:     file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib
[Info] (4456) BG:     file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313
[Info] (4456) BG:     file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib/site-packages
[Info] (4456) BG:   Looking in python search path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/DLLs'
[Info] (4456) BG:   Attempting to resolve stub package using root path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/DLLs'
[Info] (4456) BG:   Attempting to resolve using root path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/DLLs'
[Info] (4456) BG:   Looking in python search path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib'
[Info] (4456) BG:   Attempting to resolve stub package using root path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib'
[Info] (4456) BG:   Attempting to resolve using root path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib'
[Info] (4456) BG:   Looking in python search path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313'
[Info] (4456) BG:   Attempting to resolve stub package using root path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313'
[Info] (4456) BG:   Attempting to resolve using root path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313'
[Info] (4456) BG:   Looking in python search path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib/site-packages'
[Info] (4456) BG:   Attempting to resolve stub package using root path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib/site-packages'
[Info] (4456) BG:   Attempting to resolve using root path 'file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib/site-packages'
[Info] (4456) BG:   Looking in bundled stubs path 'file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/bundled/stubs'
[Info] (4456) BG:   Attempting to resolve stub package using root path 'file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/bundled/stubs'
[Info] (4456) BG:   Attempting to resolve using root path 'file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/bundled/stubs'
[Info] (4456) BG:   Looking for typeshed third-party path
[Info] (4456) BG:   Looking for typeshed stubs path
[Info] (4456) BG:   Typeshed path not found
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/AppData/Local/Programs/Python/Python313/Lib/typing.py (29ms)
[Log] (4456) BG: [BG(1)]     parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/__init__.pyi [fs read 2ms] (11ms)
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/__init__.pyi ...
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/base_events.pyi [fs read 2ms] (17ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/base_events.pyi (2ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/coroutines.pyi [fs read 2ms] (2ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/coroutines.pyi (1ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/events.pyi [fs read 1ms] (8ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/events.pyi (2ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/exceptions.pyi [fs read 1ms] (1ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/exceptions.pyi (1ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/futures.pyi [fs read 1ms] (2ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/futures.pyi (0ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/locks.pyi [fs read 1ms] (3ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/locks.pyi (1ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/protocols.pyi [fs read 1ms] (1ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/protocols.pyi (1ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/queues.pyi [fs read 1ms] (2ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/queues.pyi (1ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/runners.pyi [fs read 1ms] (2ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/runners.pyi (0ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/streams.pyi [fs read 4ms] (6ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/streams.pyi (2ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/subprocess.pyi [fs read 2ms] (5ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/subprocess.pyi (1ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/tasks.pyi [fs read 2ms] (7ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/tasks.pyi (2ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/transports.pyi [fs read 1ms] (3ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/transports.pyi (1ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/threads.pyi [fs read 1ms] (2ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/threads.pyi (0ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/taskgroups.pyi [fs read 1ms] (2ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/taskgroups.pyi (0ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/timeouts.pyi [fs read 1ms] (1ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/timeouts.pyi (1ms)
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/windows_events.pyi [fs read 1ms] (3ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/windows_events.pyi (1ms)
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/asyncio/__init__.pyi (94ms)
[Log] (4456) BG: [BG(1)]     parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/collections/abc.pyi [fs read 2ms] (2ms)
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/collections/abc.pyi ...
[Log] (4456) BG: [BG(1)]       parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/_collections_abc.pyi [fs read 1ms] (2ms)
[Log] (4456) BG: [BG(1)]       binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/_collections_abc.pyi (1ms)
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/collections/abc.pyi (3ms)
[Log] (4456) BG: [BG(1)]     parsing: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/typing_extensions.pyi [fs read 2ms] (9ms)
[Log] (4456) BG: [BG(1)]     binding: file:///c%3A/Users/Dionisia/.vscode/extensions/ms-python.vscode-pylance-2025.4.1/dist/typeshed-fallback/stdlib/typing_extensions.pyi (2ms)
[Log] (4456) BG: [BG(1)]   checking: file:///d%3A/Alan/code/projects/awaitable-tools/z.py (320ms)
[Info] (4456) BG: Heap stats: total_memory_size=14250MB, total_free_size=7142MB, total_heap_size=92MB, used_heap_size=82MB, cross_worker_used_heap_size=82MB, total_physical_size=92MB, total_available_size=1978MB, heap_size_limit=2060MB
[Log] (4456) BG: [BG(1)] analyzing: file:///d%3A/Alan/code/projects/awaitable-tools/z.py (626ms)
[Log] (4456) BG: [BG(1)] indexing: file:///d%3A/Alan/code/projects/awaitable-tools/test.py ...
[Log] (4456) BG: [BG(1)]   parsing: file:///d%3A/Alan/code/projects/awaitable-tools/test.py [fs read 1ms] (7ms)
[Log] (4456) BG: [BG(1)]   binding: file:///d%3A/Alan/code/projects/awaitable-tools/test.py (1ms)
[Log] (4456) BG: [BG(1)] indexing: file:///d%3A/Alan/code/projects/awaitable-tools/test.py [found 5] (9ms)
[Log] (4456) BG: Workspace indexing done: file:///d%3A/Alan/code/projects/awaitable-tools/test.py
[Log] (4456) BG: [BG(1)] indexing: file:///d%3A/Alan/code/projects/awaitable-tools/z.py [found 3] (0ms)
[Log] (4456) BG: Workspace indexing done: file:///d%3A/Alan/code/projects/awaitable-tools/z.py
[Log] (4456) BG: Attempting to resolve using local imports: __builtins__
[Log] (4456) BG: Attempting to resolve using root path 'file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools'
[Log] (4456) BG: [BG(1)] indexing: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/__init__.py ...
[Log] (4456) BG: [BG(1)]   parsing: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/__init__.py [fs read 0ms] (2ms)
[Log] (4456) BG: [BG(1)]   binding: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/__init__.py ...
[Log] (4456) BG: [BG(1)]     parsing: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/decorators.py [fs read 0ms] (2ms)
[Log] (4456) BG: [BG(1)]     binding: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/decorators.py (0ms)
[Log] (4456) BG: [BG(1)]     parsing: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/types.py [fs read 1ms] (2ms)
[Log] (4456) BG: [BG(1)]     binding: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/types.py (0ms)
[Log] (4456) BG: [BG(1)]     parsing: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/utils.py [fs read 0ms] (1ms)
[Log] (4456) BG: [BG(1)]     binding: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/utils.py (1ms)
[Log] (4456) BG: [BG(1)]   binding: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/__init__.py (7ms)
[Log] (4456) BG: [BG(1)] indexing: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/__init__.py [found 0] (9ms)
[Log] (4456) BG: Workspace indexing done: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/__init__.py
[Log] (4456) BG: [BG(1)] indexing: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/decorators.py [found 4] (1ms)
[Log] (4456) BG: Workspace indexing done: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/decorators.py
[Log] (4456) BG: [BG(1)] indexing: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/types.py [found 12] (0ms)
[Log] (4456) BG: Workspace indexing done: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/types.py
[Log] (4456) BG: [BG(1)] indexing: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/utils.py [found 3] (0ms)
[Log] (4456) BG: Workspace indexing done: file:///d%3A/Alan/code/projects/awaitable-tools/awaitable_tools/utils.py
```
</details>

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug. I'm going to mark this as an enhancement request and consider it in that light.

There are a couple of different features you're requesting here:
1. When `yield from` is applied to a coroutine (i.e. an object whose type derives from `Coroutine[Y, S, R]`) , the resulting type of the expression should come from the "yield type" (`Y`).
2. When one or more `yield from` statements appear within a generator-based coroutine and the operand of the `yield from` statement is an object whose type derives from `Coroutine[Y, S, R]`, then the return type of the function should be inferred as `AsyncGenerator[Y, S, R, Any]` rather than `AsyncGenerator[Y, Unknown, R, Any]`.

I'll note that pyright's behavior on 1 is consistent with mypy's (which infers `Any` in this case), and mypy doesn't support any return type inference.

I'll also note that pyright emits no errors when type checking the above code sample in standard type checking mode. The errors appear only when enabling strict mode. When using strict mode, you may need to do additional work (e.g. provide additional type annotations) to make your code type check.

This use case seems pretty esoteric to me, so I'm not convinced it's worth adding the extra logic to pyright to handle this case. There's also an easy workaround — simply provide an explicit return type annotation.

I'm going to reject this enhancement request for now. I'm willing to revisit this in the future if there's signal from enough other pyright users that this is a common-enough use case to justify the work to add these features. If you'd like to see this implemented, please upvote this issue.
